text/0014-configuration.md

```diff
@@ -0,0 +1,75 @@
+# Configuration
+- Start Date: 2019/09/09
+- RFC PR: https://github.com/openmrs/openmrs-rfc-frontend/pull/14
+
+## Definitions
+
+**Config files** allow implementers to customize the behavior of microfrontends
+by providing static files. Common formats include JSON, YAML, and HCL.
+
+A **config schema** is the set of expectations that a microfrontend has of
+implementation-provided config files. It says how the config file must be 
+structured, what keys it must have, and constrains what types of values are allowed.
+
+A **config library** is a code module that allows developers to define config 
+schemas, and loads config files in accordance with that schema. It may also 
+provide other tooling related to configuration.
+
+
+## Decision
+
+Microfrontend developers should use the official OpenMRS MF Config Library. 
+The Config Library should be a JavaScript module that
+*   Provides an API for modules to define a config schema
+*   Provides an API for modules to access config values
+*   Provides an API for implementations to provide config files
```

Comment:
👍 yep I think that would be a great way of doing it. Even better would be to maybe provide a name for the config, so that we can give better debugging messages when explaining why a specific config value is the way that it is:

```js
import Config from '@openmrs/esm-config'
import { pihCoreConfig } from `@pih/core-config`
import { mexicoConfig } from './configs/mexico`

// Second argument is name. Alternatively, name could be baked into the config itself which could be better.
Config.provide(pihCoreConfig, 'pihCore')
Config.provide(mexicoConfig, 'mexico)
```

Reply:
Oh yeah, I like that a lot.

Reply:
How can implementers provide config files if they aren't overriding esm-root-config?

Perhaps esm-root-config itself should attempt to import `module-config` from the import map, ignoring if it doesn't find it? And using it as the lowest-priority config file if it does (i.e., if someone both `provide`s a config file in a root config override and in the import map, the values from the former will take priority over the latter)?

Reply:
> How can implementers provide config files if they aren't overriding esm-root-config?

Call `Config.provide()` from any other module.

Reply:
I fear we've gotten confused talking about these things because we haven't shared the same terminology. Might be easier to figure this stuff out as we implement it instead of talking in abstractions in this RFC.

I currently lean towards all the code for all the code for config to be in esm-root-config. And that there is no other module involved, no npm package, nothing. Just code inside of esm-root-config.

Reply:
> Call Config.provide() from any other module.

I mean if the implementer isn't writing any code at all. Just throwing together modules and an import map.

> I currently lean towards all the code for all the code for config to be in esm-root-config. And that there is no other module involved, no npm package, nothing. Just code inside of esm-root-config.

Sounds great, I'll do that.

> I fear we've gotten confused talking about these things because we haven't shared the same terminology. Might be easier to figure this stuff out as we implement it instead of talking in abstractions in this RFC.

I agree. Looking forward to sharing the config library once I make some real progress with it.

Reply:
I like the idea of modules having flexibility of calling config.provide(). 
I don't like the idea of esm-root-config modifications everytime. I prefer each module that provides a configuration to register its configuration, at the time of loadup.